ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,258 @@
+# Ledger-HD sketch (UTxO-HD v2)
+
+This document describes the result of the discussion between Ledger and
+Consensus teams on 2022-01-17 about the future steps on UTxO-HD which would not
+make sense to be called this way anymore in future versions and therefore we
+propose Ledger-HD as a replacement.
+
+Below, we outline the tables that are expected to be moved to the disk, their
+dynamics, the main computations that will leave the Ledger and some open
+questions.
+
+This is meant to be just a sketch, details are not worked out yet.
+
+## Scope of Ledger-HD
+
+The plan for Ledger-HD is to move the following tables to the disk (we will show
+the "lenses" that reach each data structure from the new epoch state):
+
+- The unified map of `rewards`, `delegations`, `pointers` and `deposits`
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs     :: NewEpochState      era  -> EpochState         era  )
+  & ( esLState  :: EpochState         era  -> LedgerState        era  )
+  & ( lsDPState :: LedgerState        era  -> DPState (EraCrypto era) )
+  & ( dpsDState :: DPState (EraCrypto era) -> DState  (EraCrypto era) )
+  & ( dsUnified :: DState  (EraCrypto era) -> UMap    (EraCrypto era) )
+
+data UMap c = UMap !(Map (Credential 'Staking c) (Trip c)) !(Map Ptr (Credential 'Staking c))
+```
+
+- The current stake distribution per stake credential
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs           :: NewEpochState era -> EpochState                  era  )
+  & ( esLState        :: EpochState    era -> LedgerState                 era  )
+  & ( lsUTxOState     :: LedgerState   era -> UTxOState                   era  )
+  & ( utxosStakeDistr :: UTxOState     era -> IncrementalStake (EraCrypto era) )
+
+data IncrementalStake c = IStake
+  { credMap :: !(Map (Credential 'Staking c) Coin)
+  , ptrMap  :: !(Map Ptr Coin)
+  }
+```
+
+- The stake snapshots
+
+```haskell
+( ne :: NewEpochState era )
+  & ( nesEs       :: NewEpochState era -> EpochState           era  )
+  & ( esSnapshots :: EpochState    era -> SnapShots (EraCrypto era) )
+
+data SnapShots c = SnapShots
+  { ssStakeMark          :: SnapShot c  -- Lazy on purpose
+  , ssStakeMarkPoolDistr :: PoolDistr c -- Lazy on purpose
+  , ssStakeSet           :: !(SnapShot c)
+  , ssStakeGo            :: !(SnapShot c)
+  , ssFee                :: !Coin
+  }
+
+data SnapShot c = SnapShot
+  { ssStake       :: !(Stake c)
+  , ssDelegations :: !(VMap VB VB (Credential 'Staking c) (KeyHash 'StakePool c))
+  , ssPoolParams  :: !(VMap VB VB (KeyHash 'StakePool c) (PoolParams c))
+  }
+
+newtype Stake c = Stake
+  { unStake :: VMap VB VP (Credential 'Staking c) (CompactForm Coin)
+  }
+
+newtype PoolDistr c = PoolDistr
+  { unPoolDistr ::
+      Map (KeyHash 'StakePool c) (IndividualPoolStake c)
+  }
```

Comment:
Here's the funny thing, though, @dcoutts : once the large structures are moved to a database, ledger doesn't even need the per-pool stake distribution! it is _only_ used by consensus for the leader check, and it entirely derived from the other large data structures.

Reply:
Ok, but that doesn't obviously mean that it should be moved across the ledger/consensus boundary. I'm not saying it shouldn't but we should be clear about the reasons.

The way I would look at this is that the on-disk tables still belong to the ledger. The consensus provides the ledger with the facility to have on-disk tables (just as it provides state persistence, rollbacks etc). So something being in an on-disk table doesn't itself imply it has moved across the consensus/ledger boundary. It's like the consensus layer is the interpreter and the ledger is the logic or program being interpreted. The consensus layer shouldn't need to know about the content or purpose of the tables, it should just manage them as instructed by the ledger.

Reply:
that's totally reasonable, but we are in the ~gray~ grey zone here for the consensus/ledger split. The "ledger protocol" logic/rules were moved to the consensus repo at the vasil hard fork. and all the per-pool stake distribution really is is an aggregation of the table that we know is not up for debate.

Reply:
For this particular case, of the "protocol" rules needing access to the stake distribution, I think that does exactly fall either side of the boundary.

At a high level we would say that the Praos protocol (which is part of consensus) depends on the ledger for access to the stake distribution (or some suitable notion of a stake distribution). So maintaining that stake distribution (updating it as transactions are validated) should be the responsibility of the ledger, and then it needs to provide it to the Praos layer when needed.

So I think in this case, where one would logically draw the boundary at the high level tells us fairly clearly where we should draw the boundary here.

Reply:
Let me try this from another angle. The ledger should never be given the entire per-stake-credential stake distribution (henceforth the PSCSD :grin: ), since that violates the rule that the ledger never hold "large" structures in memory. Instead, the ledger should give consensus instructions for what to do.

The per-pool stake distribution (PPSD) is small, and so the ledger is fine to hold it. But how does it the ledger give consensus instructions on how to make it? The instructions are the delegation map, which could be provided incrementally (each delegation is a "delta", as is each pool retirement"). And this is already a part of the plan, we know that the delegation relation will be a table (probably as a part of the "unified map").

So I can see two ways to get the PPSD to the ledger:
1. it maintains it's own incremental computation, using the txins, txouts, delegations, and retirements. and hopes that this always stays in since with the PSCSD maintained by consensus using the tables and the deltas provided by ledger.
2. consensus uses the PSCSD and the delegation relation to maintain another table for the small PPSD.

And just remember, that the PPSD is not used by any code in the ledger repo any more, it is only used for the leader check. It is not even a member of what consensus deems "the ledger state".

Reply:
Yes, the reasonable designs to consider are continuously incrementally maintaining the by-pool distribution or doing a bulk computation over some phase of the epoch (implemented as an incremental traversal of the snapshot of the previous epoch by-stake-cred distribution).

Continuously incrementally maintaining the by-pool distribution sounds preferable, but also sounds tricky in the presence of fractional delegation. But if you think we can do it, then it means it's all in memory which is a lot easier.

Reply:
> And just remember, that the PPSD is not used by any code in the ledger repo any more

@dcoutts as soon as I make a bold claim things begin too change. :laughing:  SPOs will get a stake-based vote in the Voltaire phase.